lib/pwnlib/reg_sort.rb

```diff
@@ -0,0 +1,331 @@
+# encoding: ASCII-8BIT
+
+require 'pwnlib/context'
+
+module Pwnlib
+  # Topographical sort
+  module RegSort
+    # @note Do not create and call instance method here. Instead, call module method on {RegSort}.
+    module ClassMethod
+      # Walk down the assignment list of a register,
+      # return the path walked if it is encountered again.
+      #
+      # @return [Array]
+      #   The registers that involved in the cycle start from `reg`.
+      #   Empty array will be returned if no cycle start and end at `reg`.
+      # @example
+      #   check_cycle('a', {a: 1}) #=> []
+      #   check_cycle('a', {a: 'a'}) #=> ['a']
+      #   check_cycle('a', {a: 'b', b: 'c', c: 'b', d: 'a'}) #=> []
+      #   check_cycle('a', {a: 'b', b: 'c', c: 'd', d: 'a'})
+      #   #=> ['a', 'b', 'c', 'd']
+      def check_cycle(reg, assignments)
+        check_cycle_(reg, assignments.map { |k, v| [k.to_s, v] }.to_h, [])
+      end
+
+      def check_cycle_(reg, assignments, path) # :nodoc:
+        target = assignments[reg]
+        path << reg
+        # No cycle, some other value (e.g. 1)
+        return [] unless assignments.key?(target)
+        # Found a cycle
+        return target == path[0] ? path : [] if path.include?(target)
+        check_cycle_(target, assignments, path)
+      end
+
+      # Return a list of all registers which directly
+      # depend on the specified register.
+      # @example
+      #  extract_dependencies('a', {a: 1})
+      #  => []
+      #  extract_dependencies('a', {a: 'b', b: 1})
+      #  => []
+      #  extract_dependencies('a', {a: 1, b: 'a'})
+      #  => ['b']
+      #  extract_dependencies('a', {a: 1, b: 'a', c: 'a'})
+      #  => ['b', 'c']
+      def extract_dependencies(reg, assignments)
+        # .sort is only for determinism
+        assignments.select { |_, v| v == reg }.keys.map(&:to_s).sort
+      end
+
+      # Resolve the order of all dependencies starting at a given register.
+      #
+      # @example
+      #   deps = {a: [], b: [], c: ['b'], d: ['c', 'x'], x: []}
+      #   resolve_order('a', deps)
+      #   => ['a']
+      #   resolve_order('b', deps)
+      #   => ['b']
+      #   resolve_order('c', deps)
+      #   => ['b', 'c']
+      #   resolve_order('d', deps)
+      #   => ['b', 'c', 'x', 'd']
+      def resolve_order(reg, deps)
+        resolve_order_(reg, deps.map { |k, v| [k.to_s, v] }.to_h)
```

Comment:
Put `deps.map { |k, v| [k.to_s, v] }.to_h` in some kind of normalize-hash util function?